src/index.ts

```diff
@@ -0,0 +1,518 @@
+
+      possibleLocales.push(...subsets);
+    }
+
+    const matchedLocale =
+      possibleLocales
+        // Swap any locales that have aliases with their alias.
+        .map((l) => this.aliases[l] || l)
+        .find((l) => this.hasTranslationsForLocale(l));
+
+    if (matchedLocale !== this.locale) {
+      this.locale = matchedLocale;
+      this.callSubscribers();
+    }
+  }
+
+  /**
+   * Sets a rule for pluralizing a locale.
+   *
+   * @param locale  Locale or array of locales to set the rule for.
+   * @param rule    Callback that takes a number and returns a PluralCategory. Will be used whenever
+   *                pluralization is required for a key that is using the specified locale.
+   */
+  public setRule(locale: string | string[], rule: Rule): void {
+    const locales: string[] = this.normalizeLocales(locale);
+    const mapper = (l: string) => {
+      if (rule !== this.rules[l]) {
+        this.rules[l] = rule;
+        return true;
+      } else {
+        return false;
+      }
+    };
+
+    const hasChanges = locales.map(mapper).includes(true);
+
+    if (hasChanges) {
+      this.callSubscribers();
+    }
+  }
+
+  /**
+   * Sets the translations for a particular locale or locales.
+   *
+   * @param locale        Locale or list of locales that the translations should be linked to.
+   * @param translations  Nested JSON object, where object values are treated as additional context
+   *                      and string values are treated as translations. Other values are not valid.
+   */
+  public setTranslations(locale: string | string[], translations: Translations): void {
+    const locales = this.normalizeLocales(locale);
+    const clonedTranslations = this.cloneTranslations(translations);
+
+    locales.forEach((l) => {
+      this.translations[l] = clonedTranslations;
+    });
+
+    this.callSubscribers();
+  }
+
+  /**
+   * Subscribers are called when any of the internal state changes.
+   *
+   * @param handler Callback method for whenever a change occurs in the code.
+   * @returns       A method that when called, will unsubscribe the handler.
+   */
+  public subscribe(handler: Function): Unsubscribe {
+    this.subscribers.push(handler);
+
+    return () => {
+      const index: number = this.subscribers.indexOf(handler);
+      this.subscribers.splice(index, 1);
+    };
+  }
+
+  /**
+   * Find an appropriate string for a given key. This function will attempt to find the string
+   * using the primary locale, then will use the fallback if there is no match. If there are no
+   * matches, then the key will be returned.
+   *
+   * @param key   The key for looking up the string. Context should be delimited with a period.
+   * @param data  Contains data to be interpolated. `count` is a magic value for pluralization.
+   * @returns     Best string match for key given.
+   */
+  public t(key: string, data: TDataObject = {}, options: TOptions = {}): string {
```

Comment:
It's not urgent, but I think it might be nice to separate the functionality of the `t` method out into a separate, completely stateless utility. `t()` would probably only need to take a few extra arguments to make it independent of the state that `Mogna` objects hold. Since `t()` is one of the longest functions in the class, moving it to an independent module would probably make this file a lot less long, and feel a little more decoupled.

Reply:
I thought about doing that but the parameters for that call would be kind of ridiculous. Plus I want to avoid having to look up the best locale everytime you call `t`.

Something I think would be nice is maybe if we had a `TranslationLookup` interface which contained a `t` method on it. `Mnoga` would extend this interface, but it would also have a method that allows you to create a `TranslationLookup` from another method. You implemented something similar before and I think we want to do it:

For example:

```
const mnoga = new Mnoga();

// Set up translations.
mnoga.setLocale('ja');
mnoga.t('some.label');

const portuguese = mnoga.withLocale('pt');
portguese.t('some.label');
```

We're kind of an edge case because of our lessons, where we need to support two locales at once. I think we could get rid of `options` by either creating two instances of the localization services or adding a method like this. I didn't want to get too deep into it because what we have works for us. 

Thoughts?